uat/driver/utils.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+import subprocess
+
+import tenacity
+from lightkube import Client
+from lightkube.resources.batch_v1 import Job
+from lightkube.resources.core_v1 import Namespace
+
+log = logging.getLogger(__name__)
+
+
+@tenacity.retry(
+    wait=tenacity.wait_exponential(multiplier=2, min=1, max=10),
+    stop=tenacity.stop_after_attempt(30),
+    reraise=True,
+)
+def assert_namespace_active(
+    client: Client,
+    namespace: str,
+):
+    """Test that the provided namespace is Active.
+
+    Retries multiple times to allow for the K8s namespace to be created and reach Active status.
+    """
+    # raises a 404 ApiError if the namespace doesn't exist
+    ns = client.get(Namespace, namespace)
+    phase = ns.status.phase
+
+    log.info(f"Waiting for namespace {namespace} to become 'Active': phase == {phase}")
+    assert phase == "Active", f"Waited too long for namespace {namespace}!"
+
+
+def _log_before_sleep(retry_state):
+    """Custom callback to log the number of seconds before the next attempt."""
+    next_attempt = retry_state.attempt_number
+    delay = retry_state.next_action.sleep
+    log.info(f"Retrying in {int(delay)} seconds (attempts: {next_attempt})")
+
+
+@tenacity.retry(
+    wait=tenacity.wait_exponential(multiplier=2, min=1, max=64),
+    retry=tenacity.retry_if_not_result(lambda result: result),
+    before_sleep=_log_before_sleep,
+    reraise=True,
+)
+def wait_for_job(
+    client: Client,
+    job_name: str,
+    namespace: str,
+):
+    """Wait for a Kubernetes Job to complete."""
+    # raises a 404 ApiError if the Job doesn't exist
+    job = client.get(Job, name=job_name, namespace=namespace)
+    if job.status.succeeded:
+        # stop retrying, Job succeeded
+        log.info(f"Job {namespace}/{job_name} completed successfully!")
+        return True
+    elif job.status.failed:
+        raise ValueError(f"Job {namespace}/{job_name} failed!")
```

Comment:
It `job` is failed, is there a k8s mechanism that will retry it?  As is, I think the `tenacity.retry` will see this `ValueError` and just keep re-polling, rather than exiting as a failed job right away

Reply:
> It job is failed, is there a k8s mechanism that will retry it?

We've currently disabled these to ensure that if the tests fail we don't ever retry. `restartPolicy` and `backoffLimit` are the relevant options.

Reply:
Ok, so I checked this and it looks like the exception raised on failure doesn't cause tenacity to retry. It is raised normally and the execution is interrupted as intended!

Reply:
Yeah I misunderstood how this worked last week.  lgtm now.  It doesn't intercept any Exceptions but rather just looks for True/False returns.